src/ripple/json/MultivarJson.h

```diff
@@ -75,75 +107,222 @@ struct MultivarJson
 
         return (count == 0 ? none : (count < size ? some : all));
     }
-};
 
-// Wrapper for Json for all supported API versions.
-using MultiApiJson = MultivarJson<3>;
+    static constexpr struct visitor_t final
+    {
+        // Mutable Json, integral_constant version
+        template <unsigned int Version, typename... Args, typename Fn>
+        auto
+        operator()(
+            MultivarJson& json,
+            std::integral_constant<unsigned int, Version> const version,
+            Fn fn,
+            Args&&... args) const
+            -> std::invoke_result_t<
+                Fn,
+                Json::Value&,
+                std::integral_constant<unsigned int, Version>,
+                Args&&...> requires requires()
+        {
+            fn(json.val[index(Version)], version, std::forward<Args>(args)...);
+        }
+        {
+            static_assert(
+                valid(Version) && index(Version) >= 0 && index(Version) < size);
+            return fn(
+                json.val[index(Version)], version, std::forward<Args>(args)...);
+        }
 
-/*
+        // Mutable Json only, integral_constant version
+        template <unsigned int Version, typename Fn>
+        auto
+        operator()(
+            MultivarJson& json,
+            std::integral_constant<unsigned int, Version> const,
+            Fn fn) const
+            -> std::invoke_result_t<Fn, Json::Value&> requires requires()
+        {
+            fn(json.val[index(Version)]);
+        }
+        {
+            static_assert(
+                valid(Version) && index(Version) >= 0 && index(Version) < size);
+            return fn(json.val[index(Version)]);
+        }
 
-NOTE:
+        // Immutable Json, integral_constant version
+        template <unsigned int Version, typename... Args, typename Fn>
+        auto
+        operator()(
+            MultivarJson const& json,
+            std::integral_constant<unsigned int, Version> const version,
+            Fn fn,
+            Args&&... args) const
+            -> std::invoke_result_t<
+                Fn,
+                Json::Value const&,
+                std::integral_constant<unsigned int, Version>,
+                Args&&...> requires requires()
+        {
+            fn(json.val[index(Version)], version, std::forward<Args>(args)...);
+        }
+        {
+            static_assert(
+                valid(Version) && index(Version) >= 0 && index(Version) < size);
+            return fn(
+                json.val[index(Version)], version, std::forward<Args>(args)...);
+        }
```

Comment:
Can we collapse these definitions by moving the type of the first function parameter to the template parameter list?

Reply:
Good idea, will do that. Will also remove `requires requires` , since the computation of the return type with `std::invoke_result_t` makes it redundant.